tests/Unit/Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.py

```diff
@@ -0,0 +1,87 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import numpy as np
+
+
+def error(face_mesh_velocity, outward_directed_normal_covector, coords,
+          interior_gamma2, time, dim):
+    return None
+
+
+def _gauss_amplitude():
+    return 0.9
+
+
+def _gauss_width():
+    return 0.6
+
+
+def _gauss_center():
+    return 0.0
+
+
+def _center(dim):
+    center = [1.1]
+    if dim > 1:
+        center.append(0.1)
+        if dim > 2:
+            center.append(-0.9)
+    return np.asarray(center)
+
+
+def _wave_vector(dim):
+    wave_vector = [0.1]
+    if dim > 1:
+        wave_vector.append(1.1)
+        if dim > 2:
+            wave_vector.append(2.1)
+    return np.asarray(wave_vector)
+
+
+def _omega(dim):
+    wave_vector = _wave_vector(dim)
+    return np.sqrt(wave_vector.dot(wave_vector))
+
+
+def _1d_u(coords, time, dim):
+    result = -_omega(dim) * time
+    for i in range(dim):
+        result += _wave_vector(dim)[i] * (coords[i] - _center(dim)[i])
+    return result
+
+
+def _profile(u):
+    return _gauss_amplitude() * np.exp(
+        -(u - _gauss_center())**2 / _gauss_width()**2)
+
+
+def _first_deriv(u):
+    return (-2.0 * _gauss_amplitude() /
+            _gauss_width()**2) * (u - _gauss_center()) * np.exp(
+                -(u - _gauss_center())**2 / _gauss_width()**2)
+
+
+def pi(face_mesh_velocity, outward_directed_normal_covector, coords,
+       interior_gamma2, time, dim):
+    return _omega(dim) * _first_deriv(_1d_u(coords, time, dim))
+
+
+def phi(face_mesh_velocity, outward_directed_normal_covector, coords,
+        interior_gamma2, time, dim):
+    result = np.empty([dim])
+    du = _first_deriv(_1d_u(coords, time, dim))
+    for i in range(dim):
+        result[i] = _wave_vector(dim)[i] * du
+    return result
+
+
+def psi(face_mesh_velocity, outward_directed_normal_covector, coords,
+        interior_gamma2, time, dim):
+    return _profile(_1d_u(coords, time, dim))
+
+
+def constraint_gamma2(face_mesh_velocity, outward_directed_normal_covector,
+                      coords, interior_gamma2, time, dim):
+    assert interior_gamma2 >= 0.0  # make sure random gamma_2 is positive
```

Comment:
I don't think I've seen this used anywhere else -- what happens with the python asserts inside pypp in a catch test?

Reply:
The following:
```
 Filters: Unit.ScalarWave.BoundaryConditions.DirichletAnalytic                                                                                                                                                                                                                    
 The following options differ from their suggested values:                                                                                                                                                                                                                        
 Traceback (most recent call last):                                                                                                                                                                                                                                               
 File "/home/nils/SpECTRE/spectre/tests/Unit/Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.py", line 86, in constraint_gamma2                                                                                                                                 
 assert interior_gamma2 < 0.0  # make sure random gamma_2 is positive                                                                                                                                                                                                             
 AssertionError                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                  
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                  
 RunTests is a Catch v2.13.3 host application.                                                                                                                                                                                                                                    
 Run with -? for options                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                  
 -------------------------------------------------------------------------------                                                                                                                                                                                                  
 Unit.ScalarWave.BoundaryConditions.DirichletAnalytic                                                                                                                                                                                                                             
 -------------------------------------------------------------------------------                                                                                                                                                                                                  
 /home/nils/SpECTRE/spectre/tests/Unit/Evolution/Systems/ScalarWave/BoundaryConditions/Test_DirichletAnalytic.cpp:98                                                                                                                                                              
 ...............................................................................                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                  
 /home/nils/SpECTRE/spectre/tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryConditions.hpp:463: FAILED:                                                                                                                                                                
 REQUIRE( false )                                                                                                                                                                                                                                                                 
 with messages:                                                                                                                                                                                                                                                                   
 Dim := 1                                                                                                                                                                                                                                                                         
 Seed is: 697812887 from /home/nils/SpECTRE/spectre/tests/Unit/Evolution/                                                                                                                                                                                                         
 Systems/ScalarWave/BoundaryConditions/Test_DirichletAnalytic.cpp:68                                                                                                                                                                                                              
 FaceDim := 0                                                                                                                                                                                                                                                                     
 python_module := "DirichletAnalytic"                                                                                                                                                                                                                                             
 face_points := ()                                                                                                                                                                                                                                                                
 use_moving_mesh := false                                                                                                                                                                                                                                                         
 epsilon := 0.000000000001                                                                                                                                                                                                                                                        
 pretty_type::short_name<BoundaryCorrection>() := "UpwindPenalty"                                                                                                                                                                                                                 
 python_error_msg_function := "error"                                                                                                                                                                                                                                             
 python_error_message.value_or("") := ""                                                                                                                                                                                                                                          
 error_msg.value_or("") := ""                                                                                                                                                                                                                                                     
 python_tag_function := "constraint_gamma2"                                                                                                                                                                                                                                       
 pretty_type::short_name<BoundaryCorrectionTag>() := "ConstraintGamma2"                                                                                                                                                                                                           
   Failed python call with 'Function returned null'
```
I also tried explicitly raising an exception and it gives basically the same thing. We _might_ be able to improve pypp to put the Python exception message into the C++ exception message

Reply:
heh, alright not the best, but it does at least fail the test and give something that points back to what causes the error.